src/service.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent pebble service module."""
+
+import logging
+import os
+import pwd
+import time
+from pathlib import Path
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+from jinja2 import Template
+
+from charm_state import State
+
+logger = logging.getLogger(__name__)
+AGENT_SERVICE_NAME = "jenkins-agent"
+AGENT_PACKAGE_NAME = "jenkins-agent"
+SYSTEMD_SERVICE_CONF_DIR = "/etc/systemd/system/jenkins-agent.service.d/"
+PPA_GPG_KEY_ID = "67393A94A577DC24"
+READINESS_CHECK_DELAY = 30
+
+
+class PackageInstallError(Exception):
+    """Exception raised when package installation fails."""
+
+
+class ServiceRestartError(Exception):
+    """Exception raised when failing to start the agent service."""
+
+
+class JenkinsAgentService:
+    """Jenkins agent service class.
+
+    Attrs:
+       is_active: Indicate if the agent service is active and running.
+    """
+
+    def __init__(self, state: State):
+        """Initialize the jenkins agent service.
+
+        Args:
+            state: The Jenkins agent state.
+        """
+        self.state = state
+
+    def _render_file(self, path: str, content: str, mode: int) -> None:
+        """Write a content rendered from a template to a file.
+
+        Args:
+            path: the path to the file.
+            content: the data to be written to the file.
+            mode: access permission mask applied to the
+              file using chmod (e.g. 0o640).
+        """
+        with open(path, "w+", encoding="utf-8") as file:
+            file.write(content)
```

Comment:
What about [pathlib.write_text](https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_text)?

Reply:
Thanks, I updated it to use `write_text` for simplicity